File: scode/test/prob8/main.cpp
#include<cstdio>
const int N=1000010,P=1000000007;
int n,i,j,x,ans,po[N],a[N],f[N];
int main()
{
  scanf("%d",&n);
  for(po[0]=i=1;i<=n;i++)po[i]=2*po[i-1]%P;
  while(n--)scanf("%d",&x),a[x]++;
  for(i=N-1;i>1;i--){
    for(j=i,x=0;j<N;j+=i)x+=a[j];
    if(x){
      f[i]=1LL*x*po[x-1]%P;
      for(j=i+i;j<N;j+=i)f[i]=(f[i]-f[j]+P)%P;
      ans=(1LL*f[i]*i+ans)%P;
    }
  }
  printf("%d",ans);
}

File: scode/test/prob8/main.py
n=int(input())
num=[0 for i in range(1000000+1)]
mx=0
for i in input().split(' '):
	i=int(i)
	mx=max(mx,i)
	num[i]+=1
f=[0 for i in range(mx+1)]
p=[0 for i in range(mx+1)]
mod=1e9+7
p[0]=1
for i in range(1,mx+1):
	p[i]=2*p[i-1]%mod
ans=0
for i in range(mx,1,-1):
	m=sum(num[i::i])
	if m:
		f[i]=(m*p[m-1]%mod-sum(f[i::i]))%mod
		ans=(ans+f[i]*i)%mod
print(int(ans%mod))